README.md
Solr Log Manager
============

This is an implementation of a LogStash output plugin that delivers log data to Solr.  


### QUICKSTART


1. Copy the logstash_deploy folder to your disk. This contains a working version of LogStash and you can run LogStash and the Solr Output Writer from this directory.

2. Copy provided ManagedIndexSchemaFactory configured logstash_logs directory to your Solr collection folder.  If you rename the folder, also appropriately modify the core.properties.

3. If you are on a older version of Solr (before 4.4), go to the Solr Admin Page and add a new core.  In the name and instanceDir fields enter: logstash_logs. Accept the defaults for the remaining fields and click on the 'Add Core' button.

4. For a start-up example, we will use syslogs:
   Open lw_solr_syslog.conf in a text editor.  In the 'input' section of the config file add the path to your log file[s] by changing path => [ "/var/log/system.log" ]
If your Solr instance is not running on the default localhost:8983 then change the collection_host and collection_port values in the 
   ouptut section of the config file.

5. For the Sylog example, execute the following curl command (making sure to change the port)
curl 'http://localhost:8983/solr/logstash_logs/schema/fields' -X POST -H 'Content-type: application/json' --data-binary '
[{"name":"syslog_program","type":"string","indexed":true,"stored":true}]'
This will make sure you are faceting correctly on the program field.
 
6. Execute LogStash: java -jar logstash-1.3.3-flatjar.jar agent -f lw_solr_syslog.conf -p .

Log data should be displaying in your shell window and it should be getting stored in Solr.

Happy Logging!


### More Information

The MANUAL.txt file located in the logstash_deploy folder contains more details about configuring and running LogStash and writing to Solr.   

To build the lucidworks.jar yourself from within the src folder: 

	javac *.java
	jar cvf lucidworks.jar *.class
	


logstash_deploy/CHANGELOG.TXT
1.0.0 (March 27, 2014)
  # General
  -  With the exception of the @timestamp and @version fields [see below] the application will no longer try and 
  	explicitly create fields dynamically.  Collections must either use a managed_schema or an appropriately configured 
  	unmanaged schema that predefines all expected fields. 
    
    With managed schema the type of new fields will be automatically determined by Solr.    
    With unmanaged schema the user explicitly defines the type of all incoming fields ('*' can be defined in an 
    unmanaged schema to be the default type of all incoming unknown fields.) 
    
    For the mandatory fields @timestamp and @version the program will check for their existence at startup and if these fields 
    do not exist then it will attempt to create them with the following types: 

	Timestamp
		type: tdate
		name: appends prefix value from config file or logstash_ by default to 'timestamp'. ex: logstash_timestamp
		stored: true
		indexed: true
	
	Version
		type: long
		name: appends prefix value from config file or logstash_ by default to 'version'.  ex: logstash_version
		stored: true
		indexed: true

	An exception gets thrown if these fields do not exist and field creation fails.  
	
  - Uses Logstash version 1.3.3
  - Added LucidWorks Logstash 1.3.3 output file lucidworks_solr_lsv133.rb
  - Added new configuration settings that allow user better control of performance characteristics.

    output {
		  lucidworks_solr_lsv133 {
		    ...
			force_commit => ... # boolean (optional), default: false
		    flush_size => ... # number (optional), default: 100
		    idle_flush_time => ... # number (optional), default: 1
		  }
	}
    
	force_commit
		Value type is boolean
		Default is false
		
	If true then a commit request will be sent to Solr for each batch of documents.  If false then 
	the documents will be commited per the Solr instance's configured commit policy.
	
	The output uses Logstash's stud buffer to handle buffering events for batched document uploads.  The next two 
	field values get passed to the buffer manager.
	
	flush_size
	  Value type is number
	  Default is 100
	  
	Number of events to queue up before writing to Solr.  The implementation uses Logstash's stud event buffering.
	
	idle_flush_time
	  Value type is number
	  Default is 1
	  
	Amount of time in seconds since the last flush before a flush is done even if the number of buffered events is smaller than flush_size

src/CollectionManager.java
/*
 * Licensed to LucidWorks under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. LucidWorks licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class CollectionManager {
	/**
	 * 
	 * @return unique session id.
	 */
	protected String genId() {
		return UUID.randomUUID().toString();
	}
	
	protected String checkURLEncode(String data) {
		// Escape or translate invalid characters
		String result = data;
	    result = replace("<", "&lt;", result);
	    result = replace(">", "&gt;", result);
	    result = replace("@", "%40", result);
	    result = replace("%", "%25", result);
	    result = replace("&", "%26", result);
	    return result;
	}
	
	protected String replace(String target, String replacement, String data) {
	    Pattern pattern = Pattern.compile(target);
	    Matcher match = pattern.matcher(data); 
	    return match.replaceAll(replacement);
	}
}


src/LWSolrLogCollectionManager.java
/*
 * Licensed to LucidWorks under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. LucidWorks licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

/**
 * LWSolrLogCollectionManager
 * @author  MH
 * @version 1.1 
 * @since 2013 
*/

public class LWSolrLogCollectionManager extends CollectionManager{
	private String fieldsPath = "";
	private String addNewDocPath = "";
	
	private enum SOLR_RSP {
		NOCONTENT(204), NOTFOUND(404), DOCUMENTADDED(200), FIELDCREATED(200), COLLECTIONCREATED(201);

		private final int id;
		SOLR_RSP(int id) { this.id = id; }
		public int getValue() { return id; }
	}

	public void init(String host, int pt, String collection, boolean forceCommit) {
		String port = String.valueOf(pt);
		String update = "/update";
		
		if (forceCommit)
			update = "/update/?commit=true";
		
		// Note the Solr documentation examples do not include the collection.  But if
		// if collection is left out then new records are not added to the index because 
		// the fields are not found even though they do somehow appear in the managed-schema file.
		fieldsPath = "http://" + host + ":" + port + "/solr/" + collection + "/schema/fields";
		addNewDocPath = "http://" + host + ":" + port + "/solr/" + collection + update;
	}

	/**
	 * Add new field to Solr schema if field does not already exist.
	 * @param key the field name
	 * @param val the field creation string.
	 * @throws Exception
	 */
	public void createSchemaField(String key, String val) throws Exception {
		HttpURLConnection conn = null;
		try {
			// Solr will throw an exception if you try to explicitly create a field that matches a configured dynamic field pattern.
			// By using 'includeDynamic=true' fields whose name matches a configured dynamic field pattern will be reported by the 
			// server as existent even if they do not yet exist literally.  Then we skip creation here and let the field be created 
			// and typed dynamically when the document is saved in another method. 
			URL url = new URL(fieldsPath + "/" + key + "?includeDynamic=true");
			
			conn = (HttpURLConnection) url.openConnection();
			conn.setRequestMethod("GET");
			conn.setRequestProperty("Accept", "application/json");

			// Create required field if it does not already exist
			int respCode = conn.getResponseCode();
			if (respCode == SOLR_RSP.NOTFOUND.getValue()) {
				//LOG.info("Attempting to create schema field - " + url.getPath() + ",  CreateString = " + val);
				
				conn.disconnect();
				
				val = "[{" + val + "}]";
				
				url = new URL(fieldsPath);
				conn = (HttpURLConnection) url.openConnection();
				conn.setDoOutput(true);
				conn.setRequestMethod("POST");
				conn.setRequestProperty("Content-Type", "application/json");
				OutputStream os = conn.getOutputStream();
				
				os.write(val.getBytes());
				os.flush();
				respCode = conn.getResponseCode();
				if (respCode != SOLR_RSP.FIELDCREATED.getValue()) {
					throw new RuntimeException("Failed to create field <" + key + "> - \n\n" + conn.getResponseMessage());
				}
			}
		} catch (Exception e) {
			throw e;
		} finally {
			if (null != conn)
				conn.disconnect();
		}
	}

	public void flushDocs(String documents) throws Exception {
		HttpURLConnection conn = null;

		try {
			URL url = new URL(addNewDocPath);
			conn = (HttpURLConnection) url.openConnection();
			conn.setDoOutput(true);
			conn.setRequestMethod("POST");
			conn.setRequestProperty("Content-Type", "text/xml");
			OutputStream os = conn.getOutputStream();
			
			// Wrap collection of documents 
			String s = "<add>" + documents + "</add>";
			os.write(s.getBytes());
			os.flush();
			int respCode = conn.getResponseCode();
			if (respCode != SOLR_RSP.DOCUMENTADDED.getValue()) {
				throw new RuntimeException("Failed to add documents (possibly need to escape/transform/remove illegal character[s] in data prior to submission) - " + s + " - \n\n" + conn.getResponseMessage());
			}
//			LOG.info("Flushed docs = " + s.toString());
		} catch (Exception e) {
			throw e;
		} finally {
			if (null != conn)
				conn.disconnect();
		}
	}
	
	public String createSolrDocument(HashMap <String, String> event) throws Exception {
		String id = "";
		StringBuffer s = new StringBuffer();
		s.append("<doc>");

		// Check for proper encoding and create field's for tags.
		for (Map.Entry<String, String> entry : event.entrySet()) {
			String key = entry.getKey().toLowerCase().trim();
			
			if ("id".equals(key)) {
				id = checkURLEncode((String)entry.getValue());
			} else if ("tags".equals(key)) {
				String[] tags = ((String)entry.getValue()).split(",");
				for (int i = 0; i < tags.length; i++)
					s.append("<field name=\"tags\">" + checkURLEncode(tags[i]) + "</field>");
			} else {
				s.append("<field name=\"" + checkURLEncode(key) + "\">" + checkURLEncode((String)entry.getValue()) + "</field>");
			}
		}
		id = id.isEmpty() ? genId() : id;
		s.append("<field name=\"id\">" + id + "</field></doc>");
		
//		LOG.info("Added doc = " + s.toString());
		return s.toString();
	}
}



logstash_deploy/logstash/outputs/lucidworks_solr_lsv122.rb
require "logstash/namespace"
require "logstash/outputs/base"
require "net/http"
require "time"
require "pry"

require "lucidworks.jar"


# LucidWorks output that pushes Logstash collected logs to Solr. 
#
# You can learn more about LucidWorks and Solr at <http://www.lucidworks.com/>
class LogStash::Outputs::LucidWorks < LogStash::Outputs::Base

  config_name "lucidworks_solr_lsv122"
  plugin_status 1
  
  # The config values are set here to default settings.  They are overridden by the 
  # logstash conf file settings.
  
  # Solr host 
  config :collection_host, :validate => :string, :default => "localhost"

  # Port (default solr port = 8983)
  config :collection_port, :validate => :number, :default => 8983
  
  # Collection name 
  config :collection_name, :validate => :string, :default => "collection1"

  # ID of predefined External connector (not used)
  config :external_connector_id, :validate => :string, :default => "-1"

  # Display collection stats (not used)
  config :collection_stats, :validate => :boolean, :default => false
  
  # Prefix will replace the @ in logstash fieldnames @timestash and @version.
  config :field_prefix, :validate => :string, :default => "logstash_"
  
	@lucidworks
	@solrFieldCreationParams
		 
  public
  def register

	  @lucidworks = Java::LWSolrLogCollectionManager.new()

		# If an incoming field does not exist in the Solr collection then addSolrDocument will attempt to create 
		# a new field entry configured per the attributes defined in solrFieldCreationParams.  
		# The format is:
		#   field name - (note that preceding and following spaces will be trimmed.)
		#   Solr field creation parameters for the named field.  Described using json format.
		#
		# Fields not described in the hash will be created with the following default Solr field definition 
		#     "[{\"type\":\"text_en\",\"name\":\"" + fieldName + "\",\"stored\":true,\"indexed\":true}]
		#
		# User organizations can add to the hashmap configurations for LogStash fields that the organization would like
		# defined differently than the default. 
		
		@solrFieldCreationParams = Hash.new
		
		# Message is indexed for searching.  If you construct individual fields for all of the 
		# meaningful data in the message you might want to set index to false to save space.
    	@solrFieldCreationParams["message"] = "[{\"type\":\"text_general\",\"name\":\"message\",\"stored\":true,\"indexed\":true}]"

		# Version is type long
		@solrFieldCreationParams[@field_prefix + "version"] = "[{\"type\":\"long\",\"name\":\"" + @field_prefix + "version" + "\",\"stored\":true,\"indexed\":true}]"
		
		# Tags are multivalued.
		@solrFieldCreationParams["tags"] = "[{\"type\":\"text_en\",\"name\":\"tags\",\"stored\":true,\"indexed\":true,\"multiValued\":true}]"
		
		# The event timestamp is typed 'tdate'
		@solrFieldCreationParams[@field_prefix + "timestamp"] = "[{\"type\":\"tdate\",\"name\":\"" + @field_prefix + "timestamp" + "\",\"stored\":true,\"indexed\":true}]"
		
		@solrFieldCreationParams = java.util.HashMap.new(@solrFieldCreationParams)
		
		@lucidworks.init(@collection_host, @collection_port, @collection_name)
  end # def register

  public
  def receive(event)
    return unless output?(event)

		# NOTES: 
		#   1) Field names must conform to Solr field naming conventions.
		#   2) The tags field is expected to be an collection of name/value pairs.  They are here joined by commas and are
		#      later stored in the collection as multiple select field items.
		#   3) If in the future there are other fields that are collection types then a new case must be added here to either treat them as
		#      tags or to break them out as individual name/value field items as is expected in the final 'else' below.     
		#   4) Each item stored in a Solr index must have a unique ID.  You can manage the ID's by adding your own ID to 
		#      solrfields collection.  If that field is not added here then addSolrDocument call will automatically generate a GUIID 
		#      that will be the record ID.  (Note that if you pass an existing ID then the associated Solr record's data will be 
		#      overwritten.
		
		solrfields = Hash.new

		lucidfields = event.to_hash
		lucidfields.each { |key,value|
			case key
			when "tags"
				solrfields["#{key}"]= value.join(",")
			when "@timestamp"		# Remove @ from field names because @ is invalid in Solr field names
				solrfields[@field_prefix + "timestamp"] = Time.parse("#{value}").iso8601
			when "@version"
				solrfields[@field_prefix + "version"] = "#{value}"
			else
				solrfields["#{key}"] = "#{value}"
			end
		}

		begin
			@lucidworks.addSolrDocument(java.util.HashMap.new(solrfields), @solrFieldCreationParams)
		rescue

		end
		
		#binding.pry
  end # def receive
end	


logstash_deploy/logstash/outputs/lucidworks_solr_lsv133.rb
require "logstash/namespace"
require "logstash/outputs/base"
require "net/http"
require "stud/buffer"
require "time"
require 'active_support'
require "pry"

require "lucidworks.jar"


# LucidWorks output that pushes Logstash collected logs to Solr. 
#
# You can learn more about LucidWorks and Solr at <http://www.lucidworks.com/>
class LogStash::Outputs::LucidWorks < LogStash::Outputs::Base
	include Stud::Buffer

  config_name "lucidworks_solr_lsv133"
  milestone 1
  
  # The config values are set here to default settings.  They are overridden by the 
  # logstash conf file settings.
  
  # Solr host 
  config :collection_host, :validate => :string, :default => "localhost"

  # Port (default solr port = 8983)
  config :collection_port, :validate => :number, :default => 8983
  
  # Collection name 
  config :collection_name, :validate => :string, :default => "collection1"

  # Prefix will replace the @ in logstash fieldnames @timestash and @version.
  config :field_prefix, :validate => :string, :default => "logstash_"
 
  # Solr solrconfig.xml can be configured to automatically commit documents after a 
  # specified amount of time or after receipt of a maximum number of documents.  For convenience 
  # allow the logger to override this setting for use in cases where the site either isn't 
  # setting the configuration or where it is desired that logs be available for review sooner than the
  # xml configuration allows.
  #
  # If false then rely on the solrconfig setting.  If true then force an immediate commit for each received document. 
  # Note that it takes this module ~twice as long to process each document when this setting is true.
  config :force_commit, :validate => :boolean, :default => false
 
  # Number of events to queue up before writing to Solr
  config :flush_size, :validate => :number, :default => 100

  # Amount of time since the last flush before a flush is done even if
  # the number of buffered events is smaller than flush_size
  config :idle_flush_time, :validate => :number, :default => 1

  @lucidworks
 
  public
  def register

    @lucidworks = Java::LWSolrLogCollectionManager.new()

    # All fields are expected to either already be defined in the collection schema.  Or, expect that a managed-schema is being used.
    # As a convenience we here try to insure that two mandatory SiLK fields exist and if they do not we will try and 
    # have Solr create them.
    #
   
    @lucidworks.init(@collection_host, @collection_port, @collection_name, @force_commit)
 
  	@lucidworks.createSchemaField(@field_prefix + "timestamp", "\"type\":\"tdate\",\"name\":\"" + @field_prefix + "timestamp" + "\",\"stored\":true,\"indexed\":true")
  	@lucidworks.createSchemaField(@field_prefix + "version", "\"type\":\"long\",\"name\":\"" + @field_prefix + "version" + "\",\"stored\":true,\"indexed\":true")
   
    buffer_initialize(
      :max_items => @flush_size,
      :max_interval => @idle_flush_time,
      :logger => @logger
    )
  end # def register

  public
  def receive(event)
    return unless output?(event)

    # NOTES: 
    #   1) Field names must conform to Solr field naming conventions.
    #   2) The tags field is expected to be an collection of name/value pairs.  They are here joined by commas and are
    #      later stored in the collection as multiple select field items.
    #   3) If in the future there are other fields that are collection types then a new case must be added here to either treat them as
    #      tags or to break them out as individual name/value field items as is expected in the final 'else' below.     
    #   4) Each item stored in a Solr index must have a unique ID.  You can manage the ID's by adding your own ID to 
    #      solrfields collection.  If that field is not added here then addSolrDocument call will automatically generate a GUIID 
    #      that will be the record ID.  (Note that if you pass an existing ID then the associated Solr record's data will be 
    #      overwritten.
    solrfields = Hash.new
    lucidfields = event.to_hash
    lucidfields.each { |key,value|
      case key
      when "tags"
        solrfields["#{key}"]= value.join(",")
      when "@timestamp"   
        # @timestamp looks like - 2014-03-25 21:48:35 -0700 
        # Solr's format is 2014-03-25T23:48:35.591 and @ is invalid in Solr field names so fix here.
        solrfields[@field_prefix + "timestamp"] = DateTime.iso8601(Time.parse("#{value}").iso8601).strftime('%Y-%m-%dT%H:%M:%S.%LZ')
      when "@version"
        solrfields[@field_prefix + "version"] = "#{value}"
      else
        solrfields["#{key}"] = "#{value}"
      end
    }
    
    begin
      s = @lucidworks.createSolrDocument(java.util.HashMap.new(solrfields))
      buffer_receive(s)
    rescue Exception => e
      puts "Exception occured constructing new solr document - " + e.message  
    end
    #binding.pry
  end # def receive
  
  def flush(events, teardown=false)
  	begin 
   		documents = "" 
    	events.each do |event|
    		documents += event
    	end

   		@lucidworks.flushDocs(documents)
 
  	rescue Exception => e
    	@logger.warn("An error occurred while flushing events: #{e.message}")
		end
  end #def flush
end 
